core/iwasm/libraries/lib-dynlink/dynlink_section_loader.c

```diff
@@ -0,0 +1,299 @@
+{
+    DynLinkSectionExportInfo *export_info =
+        &ctx->module->dynlink_sections.export_info;
+    uint32 i;
+
+    bh_assert(ctx);
+
+    read_leb_uint32(ctx->ptr, ctx->ptr_end, export_info->count);
+
+    export_info->entries =
+        wasm_runtime_malloc(sizeof(*export_info->entries) * export_info->count);
+
+    if (export_info->entries == NULL) {
+        set_error_buf(ctx, "out of memory");
+        return false;
+    }
+
+    for (i = 0; i < export_info->count; i++) {
+        if (!(export_info->entries[i].name = load_string(ctx))) {
+            goto fail;
+        }
+        read_leb_uint32(ctx->ptr, ctx->ptr_end, export_info->entries[i].flags);
+
+        LOG_VERBOSE("Load dylink export entry success: %s %" PRIu32,
+                    export_info->entries[i].name,
+                    export_info->entries[i].flags);
+    }
+
+    LOG_VERBOSE("Load dylink export section success.");
+fail:
+    if (export_info->entries != NULL) {
+        wasm_runtime_free(export_info->entries);
+        export_info->entries = NULL;
+    }
+
+    return false;
+}
+
+static bool
+load_import_info_section(DynLinkLoaderCtx *ctx)
+{
+    DynLinkSectionImportInfo *import_info =
+        &ctx->module->dynlink_sections.import_info;
+
+    bh_assert(ctx);
+
+    if (!(import_info->module = load_string(ctx))) {
+        goto fail;
+    }
+
+    if (!(import_info->field = load_string(ctx))) {
+        goto fail;
+    }
+
+    read_leb_uint32(ctx->ptr, ctx->ptr_end, import_info->flags);
+
+    LOG_VERBOSE("Load dylink import section success: %s %s %" PRIu32,
+                import_info->module, import_info->field, import_info->flags);
+
+    return true;
+
+fail:
+    return false;
+}
+
+static bool
+dynlink_load_subsections(DynLinkLoaderCtx *ctx)
+{
+    while (ctx->ptr < ctx->ptr_end) {
+        uint8 type = *ctx->ptr++;
+        uint32 subsection_size;
+
+        read_leb_uint32(ctx->ptr, ctx->ptr_end, subsection_size);
+
+        if (!check_buf(ctx->ptr, ctx->ptr_end, subsection_size, ctx)) {
+            return false;
+        }
+
+        switch (type) {
+            case WASM_DYLINK_MEM_INFO:
+                if (!load_mem_info_section(ctx)) {
+                    return false;
+                }
+                break;
+            case WASM_DYLINK_NEEDED:
+                if (!load_needed_section(ctx)) {
+                    return false;
+                }
+                break;
+            case WASM_DYLINK_EXPORT_INFO:
+                if (!load_export_info_section(ctx)) {
+                    return false;
+                }
+                break;
+            case WASM_DYLINK_IMPORT_INFO:
+                if (!load_import_info_section(ctx)) {
+                    return false;
+                }
+                break;
+            default:
+                set_error_buf(ctx, "unknown subsection type");
+                return false;
+        }
+    }
+
+    return true;
+
+fail:
+    dynlink_sections_deinit(&ctx->module->dynlink_sections);
```

Comment:
had better not deinit here since duplicated deinit may occur in wasm loader when loading failed?